File: elixir/twelve-days/twelve_days.exs
defmodule TwelveDays do

  @gifts {
    " a Partridge in a Pear Tree",
    " two Turtle Doves, and",
    " three French Hens,",
    " four Calling Birds,",
    " five Gold Rings,",
    " six Geese-a-Laying,",
    " seven Swans-a-Swimming,",
    " eight Maids-a-Milking,",
    " nine Ladies Dancing,",
    " ten Lords-a-Leaping,",
    " eleven Pipers Piping,",
    " twelve Drummers Drumming,"
  }

  @numbers {
    "first",
    "second",
    "third",
    "fourth",
    "fifth",
    "sixth",
    "seventh",
    "eighth",
    "ninth",
    "tenth",
    "eleventh",
    "twelfth"
  }

  @doc """
  Given a `number`, return the song's verse for that specific day, including
  all gifts for previous days in the same line.
  """
  @spec verse(number :: integer) :: String.t()
  def verse(number) do
    index = number - 1
    "On the #{elem(@numbers, index)} day of Christmas my true love gave to me,#{gift_list(index)}."
  end

  @doc """
  Given a `starting_verse` and an `ending_verse`, return the verses for each
  included day, one per line.
  """
  @spec verses(starting_verse :: integer, ending_verse :: integer) :: String.t()
  def verses(starting_verse, ending_verse) do
    (starting_verse..ending_verse)
    |> Enum.map(&verse/1)
    |> Enum.join("\n")
  end

  @doc """
  Sing all 12 verses, in order, one verse per line.
  """
  @spec sing():: String.t()
  def sing do
    verses(1,12)
  end

  defp gift_list(-1), do: ""
  defp gift_list(number) do
    "#{elem(@gifts, number)}#{gift_list(number-1)}"
  end

end

File: python/twelve-days/twelve_days.py
BASE = "On the {} day of Christmas my true love gave to me: "

DAY_NAMES = [
    "first",
    "second",
    "third",
    "fourth",
    "fifth",
    "sixth",
    "seventh",
    "eighth",
    "ninth",
    "tenth",
    "eleventh",
    "twelfth"
]

VERSES = [
    "a Partridge in a Pear Tree.",
    "two Turtle Doves",
    "three French Hens",
    "four Calling Birds",
    "five Gold Rings",
    "six Geese-a-Laying",
    "seven Swans-a-Swimming",
    "eight Maids-a-Milking",
    "nine Ladies Dancing",
    "ten Lords-a-Leaping",
    "eleven Pipers Piping",
    "twelve Drummers Drumming"
]

def recite(start_verse, end_verse):
    if not isinstance(start_verse, int): raise ValueError
    if not isinstance(end_verse, int): raise ValueError
    return [verse(n) for n in range(start_verse - 1, end_verse)]

def verse(n):
    middle = ", ".join(VERSES[n:0:-1]) + ", and " if n > 0 else ""
    return BASE.format(DAY_NAMES[n]) + middle + VERSES[0]
